amgcl/detail/qr.hpp

```diff
@@ -223,10 +223,10 @@ class QR {
             const int row_stride = (Order == row_major ? nmax : 1);
             const int col_stride = (Order == row_major ? 1 : m);
 
-            for(int i = 0, ii = 0; i < n; ++i, ii += row_stride + col_stride)
+            for(int i = 0, ii = 0; i < m; ++i, ii += row_stride + col_stride)
```

Comment:
I don't think I *need* this functionality in amgcl, but having a generic QR solver/factorization could become useful in the future. So if you want to contribute this, please go ahead. The only places I currently use QR decomposition are:

1. During construction of tentative prolongation [here](https://github.com/ddemidov/amgcl/blob/47ec2de975483c9ece6c65a7356f6422490ed96f/amgcl/coarsening/tentative_prolongation.hpp#L134-L202), and only in case user provides null-space vectors. In this case I only need the factors Q and R.
2. During setup of SPAI1 smoother [here](https://github.com/ddemidov/amgcl/blob/47ec2de975483c9ece6c65a7356f6422490ed96f/amgcl/relaxation/spai1.hpp#L104-L145). In this case I only need to solve a lot of small `Ax=b` systems in the least squares sense, so I don't need the factors explicitly.

Currently, if you want to solve a linear system with QR decomposition, you need to factorize the matrix explicitly with `QR::compute`, and then to solve the system with `QR::solve`. So if we want to extend the QR solve onto the case of underdetermined systems, it may be a good idea to separate the factorization and solution interfaces completely, so that `QR::solve` will implicitly perform the required factorization; either QR or LQ based on the system matrix shape.